server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceHandler.java

```diff
@@ -146,11 +146,17 @@ public synchronized void sendFiles(GetSegmentFilesRequest request, ActionListene
                 );
             };
             cancellableThreads.checkForCancel();
-            final IndexShardRoutingTable routingTable = shard.getReplicationGroup().getRoutingTable();
-            ShardRouting targetShardRouting = routingTable.getByAllocationId(request.getTargetAllocationId());
-            if (targetShardRouting == null) {
-                logger.debug("delaying replication of {} as it is not listed as assigned to target node {}", shard.shardId(), targetNode);
-                throw new DelayRecoveryException("source node does not have the shard listed in its state as allocated on the node");
+            if (shard.isPrimaryMode()) {
```

Comment:
Even with this check we have a race where it can flip out of primaryMode.  A few things we should look at here - 
1. If the shard is flipping out of primary mode, active replication events where that shard is the source should be getting cancelled.  If this were to throw with not in primary mode we can handle that as a cancellation.  Even if it were to throw today, the replicas should catch up to the newly elected primary.
2. I don't think we need this shardRouting check at all given the replica chunks are being sent to is not in recovery.  This check be handled by RecoverySourceHandler already.

Reply:
Correct, but i would leave that for a separate change.  I'm more interested in why a failure here would fail the test?

Reply:
The test fails for point 2 because of assert failure on primary mode. assert failure doesn't throw an exception, if some excception was thrown then SR event would be cancelled and relocation will be complete correctly.